### PR TITLE
[YARP] Implement compilation for Regex / InterpolatedRegex

### PR DESCRIPTION
**Pull request by jemmaissroff**

### CODE REVIEW

File: yarp/yarp_compiler.c
Comment (kddnewton) [inline yarp/yarp_compiler.c:1282]: This should be a `yp_string_node_t *`.

File: yarp/yarp_compiler.c
Comment (kddnewton) [inline yarp/yarp_compiler.c:1678]: I'm assuming the `0` there is the options. Would you like to do that in a follow-up PR?

**Review by kddnewton — CHANGES_REQUESTED** [review]

**Review by kddnewton — APPROVED** [review]
I'll leave it up to you, but you might want to merge #8395 and then rebase so you can reuse the interpolation logic.

**Comment by jemmaissroff** [comment]
> I'll leave it up to you, but you might want to merge #8395 and then rebase so you can reuse the interpolation logic.

That was my plan! There's just one flaky test that I re-tried. Once that's green I'll merge and rebase

**Review by kddnewton — APPROVED** [review]

File: yarp/yarp_compiler.c
Comment (eregon) [inline yarp/yarp_compiler.c:1658]: As a note, this will be affected by this bug: https://github.com/ruby/yarp/issues/1452
And the YP_NODE_INTERPOLATED_REGULAR_EXPRESSION_NODE case too (https://github.com/ruby/yarp/issues/1452#issuecomment-1716271367).

**Pull request merged**
